### PR TITLE
Add codes REST API

### PR DESCRIPTION
**Pull request by amolicas**
Add codes to the REST API, with next calls:
GET /api/codes
GET /api/codes/:cid
POST /api/codes
PUT /api/codes/:cid

### CODE REVIEW

File: _rest_routes.inc.php [hunk 1]
```diff
@@ -33,6 +34,7 @@
 use OpenEMR\RestControllers\ImmunizationRestController;
 use OpenEMR\RestControllers\InsuranceRestController;
 use OpenEMR\RestControllers\MessageRestController;
+use OpenEMR\RestControllers\CodesRestController;
```
Comment (stephenwaite) [inline _rest_routes.inc.php:37]: hi @amolicas , pretty trivial but keeping these sorted alphabetically, thank you

File: src/RestControllers/CodesRestController.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * @copyright Copyright (c) 2020 Alberto Moliner <amolicas79@gmail.com>
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+
+
+namespace OpenEMR\RestControllers;
+
+use OpenEMR\Services\CodesService;
+use OpenEMR\RestControllers\RestControllerHelper;
+
```
Comment (stephenwaite) [inline src/RestControllers/CodesRestController.php:17]: also sorted alpha here please

File: _rest_routes.inc.php (hunk 1 above)
Comment (amolicas) [reply _rest_routes.inc.php:37]: Hi @stephenwaite, sorry this is my first contribution, do you mean to sort this way?

use OpenEMR\RestControllers\CodesRestController;
use OpenEMR\RestControllers\DocumentRestController;
use OpenEMR\RestControllers\ImmunizationRestController;
use OpenEMR\RestControllers\InsuranceRestController;
use OpenEMR\RestControllers\MessageRestController;

Comment (stephenwaite) [reply _rest_routes.inc.php:37]: no prob, yes, looks perfect, and welcome aboard!

Comment (amolicas) [reply _rest_routes.inc.php:37]: Ok, I will commit this way, thanks!

**Comment by amolicas** [comment]
I don't know why I get this error on Travis CI build. Need help to finish the pull request.

------------------------------
Checking for PHP syntax errors
------------------------------
PHP Warning:  Private methods cannot be final as they are never overridden by other classes in ./portal/patient/fwk/libs/verysimple/Phreeze/Criteria.php on line 195
---------------------------------------
FAILED - Checking for PHP syntax errors
---------------------------------------

**Comment by sjpadgett** [comment]
The classes probably need to be in their own namespace. Same with adding OpenEMR namespaces. I'd ignore this for now and i'll look into ensuring portal classes get their own namespace.

**Comment by bradymiller** [comment]
hi @amolicas ,

The code looks like on the right track in terms of wiring for api. The main issue is what a concept of code is in OpenEMR. The codes table is only part of that.

In OpenEMR, check out, Administration->Forms->Lists->Code Types

That is a listing of potential Code Types in OpenEMR. And you'll note the column at the far right called external, which means that the Codes for that code type are not stored in the codes table, but in another table. So, in order to look up codes, OpenEMR does not directly look in the codes table, but instead uses a function in custom/code_types.inc.php that automatically figures out where to look on code look up. And writing codes is a bit more complicated; when stored in codes table then can directly modify; when not stored in codes table, then in that case can't modify the other table, but instead puts a modified code in the codes table! (don't worry the lookup functions in custom/code_types.inc.php deal with that when looking up codes).

So, to support api for codes, wouldn't really support writing new codes. And would really just have a couple GET endpoints that would likely look something like:
```
GET /api/codes/ (a getall that could search all code types via supplied search criteria)
GET /api/codes/:ct (a getall would search specific code type (those are the Key column in the Code Types list described above))
GET /api/codes/:ct/:co (a getone where are providing codetype and code)
```
Note that all forms of this searching are done in the Adminisitration->Coding->codes gui, so can see how that code work (which are basically leveraging functions from custom/code_types.inc.php).

**Comment by amolicas** [comment]
Hi @bradymiller ,
thank you very much for the explicanation, I didn't consider that complexity. Now I am using the API to add some services or medications from another system, but I don't use all code types, so I understand that it is good for me but it is not valid for all uses. I will try to modify it as you said in order to collaborate with the community.
Thanks

**Pull request closed without merging**